README
 
These files are part of the plugin gr-spi of GNU Radio for BeagleSDR (https://github.com/mhe747/dawnboard)
GNU Radio is a free & open-source software development toolkit that provides signal processing blocks to implement software radios. 
It can be used with readily-available low-cost external BeagleSDR RF hardware on top of Beagleboard X15 in order to create software-defined radios, or without hardware in a simulation-like environment. 
It is widely used in hobbyist, academic and commercial environments to support both wireless communications research and real-world radio systems. 
 
About the plugin :
 
	https://github.com/mhe747/dawnboard/tree/master/gnuradio


ChangeLog
#
# Copyright 2001,2002,2003,2004,2005,2006,2007 Free Software Foundation, Inc.
# 
# This file is part of GNU Radio
# 
# GNU Radio is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3, or (at your option)
# any later version.
# 
# GNU Radio is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with GNU Radio; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street,
# Boston, MA 02110-1301, USA.
# 

gr-spi is a plugin for spi control of the BeagleSDR, an add-on board that plug into Beagleboard X15


lib/gr_spi_source_s.cc
/* -*- c++ -*- */
/*
 * gr_spi_source_s.cc
 * George Schaertl <gschaertl@gmail.com>
 * 5 May 2009
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <spi/gr_spi_source_s.h>
#include <gnuradio/io_signature.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/types.h>
#include <linux/spi/spidev.h>

gr_spi_source_s_sptr gr_make_spi_source_s ()
{
	return gr_spi_source_s_sptr (new gr_spi_source_s ());
}

static const int MIN_IN = 0;
static const int MAX_IN = 0;
static const int MIN_OUT = 1;
static const int MAX_OUT = 1;

gr_spi_source_s::gr_spi_source_s ()
	: sync_block("spi_source_s",
	                gr::io_signature::make(MIN_IN, MAX_IN, sizeof (short)),
	                gr::io_signature::make(MIN_OUT, MAX_OUT, sizeof (short)))
{
}

gr_spi_source_s::~gr_spi_source_s ()
{
}

bool gr_spi_source_s::send_spi_command (const std::string &command)
{
	d_command_buffer.append(command);
	return true;
}

bool gr_spi_source_s::open_spi (const std::string &device, unsigned int speed)
{
	int ret;
	d_spi_speed = speed;
	d_spi_bits = 16;
	d_spi_mode = 0;
	d_command_buffer = "";

	d_fd = open (device.c_str(), O_RDWR);
	if (d_fd < 0) return false;

	ret = ioctl (d_fd, SPI_IOC_WR_MODE, &d_spi_mode);
	if (ret == -1) return false;

	ret = ioctl (d_fd, SPI_IOC_WR_BITS_PER_WORD, &d_spi_bits);
	if (ret == -1) return false;

	ret = ioctl (d_fd, SPI_IOC_WR_MAX_SPEED_HZ, &d_spi_speed);
	if (ret == -1) return false;

	return true;
}

bool gr_spi_source_s::close_spi()
{
	close(d_fd);
	return true;
}

int gr_spi_source_s::work (int noutput_items,
          gr_vector_const_void_star &input_items,
          gr_vector_void_star &output_items)
{
	short *out = (short *) output_items[0];
	int length = noutput_items * d_spi_bits / 8;
	int ret;

	unsigned char txbuf[length], rxbuf[length];

	for (int i = 0; i < length; i++) {
		if (d_command_buffer.empty()) break;
		txbuf[i]  = d_command_buffer[0];
	}

	struct spi_ioc_transfer tr;
	tr.tx_buf = (unsigned long) txbuf,
	tr.rx_buf = (unsigned long) rxbuf,
	tr.len = length,
	tr.delay_usecs = 0,
	tr.speed_hz = d_spi_speed,
	tr.bits_per_word = d_spi_bits,

	ret = ioctl(d_fd, SPI_IOC_MESSAGE(1), &tr);
	if (ret == 1) return -1;

	for (int i = 0; i < noutput_items; i++) {
		out[i]  = rxbuf[i*2] << 8;
		out[i] &= 0x0fff;
		out[i] |= rxbuf[i*2 + 1];
	}

	return noutput_items;
}


lib/gr_spi_srcsink_ss.cc
/* -*- c++ -*- */
/*
 * gr_spi_srcsink_ss.cc
 * George Schaertl <gschaertl@gmail.com>
 * 5 May 2009
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <spi/gr_spi_srcsink_ss.h>
#include <gnuradio/io_signature.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/types.h>
#include <linux/spi/spidev.h>
#include <endian.h>

#if FAKE_IOCTL
int ioctl(int fd, int req, void* arg) {
	return 0;
}
#endif

gr_spi_srcsink_ss_sptr gr_make_spi_srcsink_ss ()
{
	return gr_spi_srcsink_ss_sptr (new gr_spi_srcsink_ss ());
}

static const int MIN_IN = 1;
static const int MAX_IN = 1;
static const int MIN_OUT = 1;
static const int MAX_OUT = 1;

// The kernel doesn't like more than 4096 bytes in one spi_transfer
static const int MAX_ITEMS = 2048;

gr_spi_srcsink_ss::gr_spi_srcsink_ss ()
	: sync_block("spi_srcsink_ss",
	                gr::io_signature::make(MIN_IN, MAX_IN, sizeof (short)),
	                gr::io_signature::make(MIN_OUT, MAX_OUT, sizeof (short)))
{
	set_output_multiple(MAX_ITEMS); // Short spi_transfers are inefficient

	d_swap_bytes = (BYTE_ORDER != BIG_ENDIAN);
}

gr_spi_srcsink_ss::~gr_spi_srcsink_ss ()
{
}

bool gr_spi_srcsink_ss::open_spi (const std::string &device, unsigned int speed)
{
	int ret;
	d_spi_speed = speed;
	d_spi_bits = 16;
	d_spi_mode = 0;

	d_fd = open (device.c_str(), O_RDWR);
	if (d_fd < 0) return false;

	ret = ioctl (d_fd, SPI_IOC_WR_MODE, &d_spi_mode);
	if (ret == -1) return false;

	ret = ioctl (d_fd, SPI_IOC_WR_BITS_PER_WORD, &d_spi_bits);
	if (ret == -1) return false;

	ret = ioctl (d_fd, SPI_IOC_WR_MAX_SPEED_HZ, &d_spi_speed);
	if (ret == -1) return false;

	return true;
}

bool gr_spi_srcsink_ss::close_spi()
{
	close(d_fd);
	return true;
}

void gr_spi_srcsink_ss::swap_bytes(char* stream, int len) {
	for (len-= len% 2; len ; len -= 2) {
		char     temp = stream[len-1];
		stream[len-1] = stream[len-2];
		stream[len-2] = temp;
	}
}

int gr_spi_srcsink_ss::work (int noutput_items,
          gr_vector_const_void_star &input_items,
          gr_vector_void_star &output_items)
{
	if (noutput_items > MAX_ITEMS) noutput_items = MAX_ITEMS;

	if (d_swap_bytes) swap_bytes(
		(char*) &input_items[0],
		noutput_items*sizeof(short));

	int ret;

	struct spi_ioc_transfer tr;
	tr.tx_buf = (unsigned long) &input_items[0],
	tr.rx_buf = (unsigned long) &output_items[0],
	tr.len = noutput_items * sizeof(short),
	tr.delay_usecs = 0,
	tr.speed_hz = d_spi_speed,
	tr.bits_per_word = d_spi_bits,

	ret = ioctl(d_fd, SPI_IOC_MESSAGE(1), &tr);
	if (ret == 1) return -1;

        if (d_swap_bytes) swap_bytes(
		(char*) &output_items[0],
		noutput_items*sizeof(short));

	return noutput_items;
}


include/spi/spi_srcsink_ss.h
/* -*- c++ -*- */
/*
 * spi_srcsink_ss.h
 * George Schaertl <gschaertl@gmail.com>
 * 14 July 2009
 */

#ifndef INCLUDED_SPI_SRCSINK_SS_H
#define INCLUDED_SPI_SRCSINK_SS_H

#include <gnuradio/sync_block.h>

class spi_srcsink_ss;

typedef boost::shared_ptr<spi_srcsink_ss> spi_srcsink_ss_sptr;

spi_srcsink_ss_sptr spi_make_srcsink_ss (char *device, unsigned int speed);

class spi_srcsink_ss : public gr::sync_block
{
	friend spi_srcsink_ss_sptr spi_make_srcsink_ss (
					char *device, unsigned int speed);
	
private:
	int d_fd;
	unsigned int d_spi_speed;
	unsigned char d_spi_bits;
	unsigned char d_spi_mode;
	bool d_swap_bytes;

	void swap_bytes(char* stream, int len);

protected:
	spi_srcsink_ss (char *device, unsigned int speed);

public:
	~spi_srcsink_ss ();

	int work (int noutput_items,
	          gr_vector_const_void_star &input_items,
	          gr_vector_void_star &output_items);

};

#endif /* INCLUDED_SPI_SRCSINK_SS_H */


include/spi/gr_spi_source_s.h
/* -*- c++ -* */
/*
 * gr_spi_source_s.h
 * George Schaertl <gschaertl@gmail.com>
 * 5 May 2009
 */

#ifndef INCLUDED_GR_SPI_SOURCE_S_H
#define INCLUDED_GR_SPI_SOURCE_S_H

#include <gnuradio/sync_block.h>

class gr_spi_source_s;

typedef boost::shared_ptr<gr_spi_source_s> gr_spi_source_s_sptr;

gr_spi_source_s_sptr gr_make_spi_source_s ();

class gr_spi_source_s : public gr::sync_block
{
	int d_fd;
	unsigned int d_spi_speed;
	unsigned char d_spi_bits;
	unsigned char d_spi_mode;
	std::string d_command_buffer;

private:
	friend gr_spi_source_s_sptr gr_make_spi_source_s ();
	gr_spi_source_s ();

public:
	~gr_spi_source_s ();

	int work (int noutput_items,
	          gr_vector_const_void_star &input_items,
	          gr_vector_void_star &output_items);

	bool open_spi (const std::string &device,
	               unsigned int speed);

	bool close_spi ();

	bool send_spi_command (const std::string &command);
};

#endif /* INCLUDED_GR_SPI_SOURCE_S_H */


include/spi/gr_spi_srcsink_ss.h
/* -*- c++ -*- */
/*
 * gr_spi_srcsink_ss.h
 * George Schaertl <gschaertl@gmail.com>
 * 5 May 2009
 */

#ifndef INCLUDED_GR_SPI_SRCSINK_SS_H
#define INCLUDED_GR_SPI_SRCSINK_SS_H

#include <gnuradio/sync_block.h>

class gr_spi_srcsink_ss;

typedef boost::shared_ptr<gr_spi_srcsink_ss> gr_spi_srcsink_ss_sptr;

gr_spi_srcsink_ss_sptr gr_make_spi_srcsink_ss ();

class gr_spi_srcsink_ss : public gr::sync_block
{
	int d_fd;
	unsigned int d_spi_speed;
	unsigned char d_spi_bits;
	unsigned char d_spi_mode;
	bool d_swap_bytes;

private:
	friend gr_spi_srcsink_ss_sptr gr_make_spi_srcsink_ss ();
	gr_spi_srcsink_ss ();
	void swap_bytes(char* stream, int len);

public:
	~gr_spi_srcsink_ss ();

	int work (int noutput_items,
	          gr_vector_const_void_star &input_items,
	          gr_vector_void_star &output_items);

	bool open_spi (const std::string &device,
	               unsigned int speed);

	bool close_spi ();

};

#endif /* INCLUDED_GR_SPI_SOURCE_S_H */


lib/spi_srcsink_ss.cc
/* -*- c++ -*- */
/*
 * spi_srcsink_ss.cc
 * George Schaertl <gschaertl@gmail.com>
 * 14 July 2009
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <spi/spi_srcsink_ss.h>
#include <gnuradio/io_signature.h>
#include <fcntl.h>
#include <linux/types.h>
#include <linux/spi/spidev.h>
#include <sys/ioctl.h>
#include <endian.h>
#include <stdexcept>
#include <iostream>
#include <cstring>
#include <cerrno>

#ifdef FAKE_IOCTL
int ioctl(int fd, int req, void* arg) {
	return 0;
}
#endif

spi_srcsink_ss_sptr spi_make_srcsink_ss (char* device, unsigned int speed)
{
	return spi_srcsink_ss_sptr (new spi_srcsink_ss (device, speed));
}

static const int MIN_IN = 1;
static const int MAX_IN = 1;
static const int MIN_OUT = 1;
static const int MAX_OUT = 1;

// The kernel doesn't like more than 4096 bytes in one spi_transfer
static const int MAX_ITEMS = 4096/sizeof(short);

spi_srcsink_ss::spi_srcsink_ss (char *device, unsigned int speed)
	: sync_block("spi_srcsink_ss",
	                gr::io_signature::make(MIN_IN, MAX_IN, sizeof (short)),
	                gr::io_signature::make(MIN_OUT, MAX_OUT, sizeof (short)))
{
	set_output_multiple(MAX_ITEMS); // Short spi_transfers are inefficient

//	The BeagleBoard lies to us about this for some reason
//	d_swap_bytes = (BYTE_ORDER != BIG_ENDIAN);
	d_swap_bytes = false;

        int ret;
        d_spi_speed = speed;
        d_spi_bits = 16;
        d_spi_mode = 0;

        d_fd = open (device, O_RDWR);
        if (d_fd < 0) throw std::runtime_error("can't open device");
        
	ret = ioctl (d_fd, SPI_IOC_WR_MODE, &d_spi_mode);
        if (ret == -1) throw std::runtime_error("can't set spi mode");

        ret = ioctl (d_fd, SPI_IOC_WR_BITS_PER_WORD, &d_spi_bits);
        if (ret == -1) throw std::runtime_error("can't set spi bits per word");

        ret = ioctl (d_fd, SPI_IOC_WR_MAX_SPEED_HZ, &d_spi_speed);
        if (ret == -1) throw std::runtime_error("can't set spi max speed");

	std::cerr << "srcsink constructor exiting successfully" << std::endl;
}

spi_srcsink_ss::~spi_srcsink_ss ()
{
	close(d_fd);
}

void spi_srcsink_ss::swap_bytes(char* stream, int len) {
	for (len-= len% 2; len ; len -= 2) {
		char     temp = stream[len-1];
		stream[len-1] = stream[len-2];
		stream[len-2] = temp;
	}
}

int spi_srcsink_ss::work (int noutput_items,
          gr_vector_const_void_star &input_items,
          gr_vector_void_star &output_items)
{

	// input_items and output_items are vectors of pointers to vectors of
	// samples, one vector-of-samples for each input or output stream.
	// Forget this like I did and you get segfaults in strange places. )-:
	short* in  = (short *) input_items[0];
	short* out = (short *) output_items[0];	

	if (d_swap_bytes)
		swap_bytes( (char*) in, noutput_items*sizeof(short) );

/*
//	This *was* a good idea, but the ioctl() segfaults on the BeagleBoard.
	int n_xfers = noutput_items / MAX_ITEMS;
	struct spi_ioc_transfer *xfer = new struct spi_ioc_transfer[n_xfers];

	for (int i=0; i<n_xfers; i++) {
		int offset = i * MAX_ITEMS;
		xfer[i].tx_buf = (unsigned long) &(in[offset]);
		xfer[i].rx_buf = (unsigned long) &(out[offset]);
		xfer[i].len = MAX_ITEMS * sizeof(short);
		xfer[i].speed_hz = d_spi_speed;
		xfer[i].delay_usecs = 0;
		xfer[i].bits_per_word = d_spi_bits;
		xfer[i].cs_change = 0;
	}

	std::cerr << "Trying ioctl with n_xfers = " << n_xfers << std::endl;
	int ret = ioctl(d_fd, SPI_IOC_MESSAGE(n_xfers), xfer);
	if (ret == -1)
		throw std::runtime_error(
		 strcat("SPI transfer failed: ",strerror(errno)));

	delete xfer;
*/

	for(int offset = 0; offset < noutput_items; offset += MAX_ITEMS)
	{
		int n_items = MAX_ITEMS; // make sure we set_output_multiple()!

		struct spi_ioc_transfer tr;
		tr.tx_buf = (unsigned long) &(in[offset]);
		tr.rx_buf = (unsigned long) &(out[offset]);
		tr.len = n_items * sizeof(short);
		tr.delay_usecs = 0;
		tr.speed_hz = d_spi_speed;
		tr.bits_per_word = d_spi_bits;
		tr.cs_change = 0;

		int ret = ioctl(d_fd, SPI_IOC_MESSAGE(1), &tr);
		if (ret == -1) throw std::runtime_error(
			strcat("SPI transfer failed: ",strerror(errno)));
	}

	if (d_swap_bytes)
		swap_bytes( (char*) out, noutput_items*sizeof(short) );

	return noutput_items;
}
